Add initial networking layer

momo-ozawa opened the pull request:
## Description

Resolves https://github.com/wordpress-mobile/WordPress-iOS/issues/16863

This PR adds the initial networking layer:

- Added `InAppPurchasesService` to call the fetch product skus / create order endpoints
- Added initial `InAppPurchasesServiceTests`
- Added `InAppPurchasesAPIRouter` to organize the endpoints 
- Added Alamofire as a package dependency

## Notes
- I will add more tests to InAppPurchasesService in a later PR

## How to test 

1. Go to Edit Scheme > Run > Options > StoreKit Configuration
2. Choose `Configuration.storekit`
3. Run the iOS app
4. Tap on Rocket Fuel to purchase
5. Tap Confirm
6. Tap OK
7. On Xcode debugger menu, click on the Manage StoreKit Transactions menu item (See example below)

<img width="536" alt="Screen Shot 2021-07-19 at 16 29 19" src="https://user-images.githubusercontent.com/6711616/126186328-1bc968f4-8c18-45ad-84f1-b5033fc1cbe7.png">

8. ✅ There should be a record of Rocket Fuel being purchased (See example below)

<img width="747" alt="Screen Shot 2021-07-19 at 16 32 32" src="https://user-images.githubusercontent.com/6711616/126186582-3f76559e-7fe5-45ae-9068-2754883f6351.png">

Commit:
Add: Networking protocol

Commit:
Fix: swiftlint whitespace

Commit:
Refactor: get receipt and create order if txn state is purchased or restored

Commit:
Add: Alamofire as a package dependency

Commit:
Add: APIRequest protocol

Commit:
Fix: protocol conformance

Commit:
Refactor: rename Networking to InAppPurchasesAPIProtocol

Commit:
Add: InAppPurchases service

Commit:
Add: InAppPurchasesServiceTests

Commit:
Add: InAppPurchasesRouter

Commit:
Add: generic Alamofire request method

Commit:
Delete: APIRequest

emilylaguna commented on MobilePayKit/Sources/MobilePayKit/Services/AppStore/AppStoreService.swift:
```diff
@@ -77,22 +103,79 @@ extension AppStoreService: SKPaymentTransactionObserver {
+
+        iapService.createOrder(
+            identifier: product.productIdentifier,
+            price: Int(truncating: product.price),
```
The value of the price should be in cents, if the price is `5.99` then it should be converted to: `599`.

In theory this should work unless there's a better way: 

`Int(price.floatValue * 100)`

momo-ozawa replied:
Fixed. cd1fb1f

Thank you for updating the API docs as well! 🙇‍♀️

leandroalonso commented:
@momo-ozawa These tests are failing:

`testPaymentQueueUpdatedTransactions_WhenTransactionStateIsPurchased_CallsFinishTransaction`
`testPaymentQueueUpdatedTransactions_WhenTransactionStateIsRestored_CallsFinishTransaction`

emilylaguna commented on MobilePayKit/Sources/MobilePayKit/Services/AppStore/AppStoreService.swift:
```diff
@@ -77,22 +103,79 @@ extension AppStoreService: SKPaymentTransactionObserver {
+
+        guard let appStoreReceiptURL = Bundle.main.appStoreReceiptURL,
+              FileManager.default.fileExists(atPath: appStoreReceiptURL.path) else {
+            print("Could not find app store receipt")
```
Should we handle the returns as errors?

momo-ozawa replied:
Yes! I'm thinking of creating a separate PR for error handling. 

I think this PR might get big if I changed the CompletionCallbacks from `(T) -> Void` to something like `(Result<T, Error>) -> Void`.

emilylaguna replied:
Makes sense to me!

emilylaguna reviewed (changes requested):
Hiya! Overall this works really well! I left a couple comments below.

In addition the tests that are failing, I found the macOS example doesn't build because `navigationBarItems` is for iOS only. 

Also, I'm unsure we need to use Alamofire since the requests themselves are pretty simple, and we should be able to use URLSession's instead. Unless there was a specific reason you chose to use AF?

leandroalonso commented on MobilePay-Apple.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:
```diff
@@ -0,0 +1,16 @@
+  "object": {
+    "pins": [
+      {
+        "package": "Alamofire",
```
@momo-ozawa there's any special reason why we need `Alamofire`?

I think it would be good to keep this library without dependencies and as much light as possible. Nowadays `URLSession` is very powerful so I think we can go with it.

This can be refactored on a future PR though.

Commit:
Refactor: convert price to price in cents

leandroalonso commented:
@momo-ozawa as we talked, let's ignore the failing tests for now and then we can get back to those later.

Commit:
Refactor: skip tests for now

For the purchased / restored tests to pass, we need to build a receipt stub and a mock iapService.

momo-ozawa commented:
@leandroalonso @emilylaguna 
I added a separate task for the macOS build on the project issue.

I'll be removing Alamofire in favor of URLSession in my next PR 🙇‍♀️

emilylaguna reviewed (approved):
LGTM! 🚀

momo-ozawa commented:
Closing this PR due to merge conflicts -- I will create a new PR!
